### DISCUSSION TITLE
Is it possible to compile in multiple network drivers, and switch at runtime?

### DISCUSSION THREAD

**Discussion by victorallume** [micropython/micropython#9459] [Libraries & Drivers] [answered]
I'm wanting a particular board to be able to switch between wifi and ethernet. I've tried this on both a Pico W and ESP32 board, both connected to wiznet5k ethernet chips, but after flashing, when I import `network` it only has `WLAN` or `WIZNET`, not both. To compile, I've tried:

-  PICO W with wiznet 5100 hat: `make BOARD=PICO_W MICROPY_PY_NETWORK_WIZNET5K=W5100S`, this only has `network.WLAN` available, no WIZNET
-  PICO W with wiznet 5100 hat: `make BOARD=W5100S_EVB_PICO MICROPY_PY_NETWORK_CYW43=1`, this only has `network.WIZNET` available, no WLAN
- similar results in the ESP32.

Any pointers?

**Comment by robert-hh** [comment]
Are you sure the the first two items of the bullet list are right? The make commands are identical.

**Reply by victorallume** [reply]
Thanks; have updated the make command

**Reply by robert-hh** [reply]
Almost right, I assume.

**Reply by victorallume** [reply]
my failure in copy-pasta :-/

**Comment by jimmo** [comment, marked as answer]
`make BOARD=PICO_W MICROPY_PY_NETWORK_WIZNET5K=W5100S` (and your second command to do the reverse configuration) doesn't do what you expect on the rp2 port because these variables are not understood by make. Rather you need to set them for CMake.

To get this working, you should modify `ports/rp2/boards/PICO_W/mpconfigboard.cmake` to add

```
set(MICROPY_PY_NETWORK_WIZNET5K W5100S)
```

Or equivalently add

```
set(MICROPY_PY_NETWORK_CYW43 ON)
```

to `ports/rp2/boards/W5100S_EVB_PICO/mpconfigboard.cmake`.

We should probably add a `CMAKE_ARGS_EXTRA` option to `ports/rp2/Makefile`

**Reply by victorallume** [reply]
Thanks; that worked; I can even get both wifi and ethernet connected at the same time. Out of interest, how does `socket` decide which interface to use if there are multiple?

**Reply by jimmo** [reply]
You use `bind()` to bind the socket interface (by its address).
